### ISSUE TITLE
apt plugin creates incorrect class path in IntelliJ

### ISSUE THREAD

**Issue by mdeinum** [tbroyer/gradle-apt-plugin#58]
When using the 0.9 version of the plugin we have a class path with a dependency set to the scope **Compile**. 

However when upgrading (both 0.10 and 0.11) and adding the `net.ltgt.apt-idea` plugin that dependency suddenly changes to be scoped **Provided**. As IntelliJ only places the scopes **Runtime** and **Compile** on the class path and skips **Provided** this leads to startup issues. 

Judging from our dependencies it should have been **Compile** and not **Provided**. Looks like the plugin is resulting in an incorrect class path in IntelliJ.

Using Gradle 4.1.

**Comment by tbroyer** [comment]
In v0.10 (and v0.11), the plugin tries to align with what the Gradle version does, which changed in Gradle 3.4: https://github.com/tbroyer/gradle-apt-plugin/commit/16b967244f9356649d817fd7c9457464b3990271#diff-872e6b905b5ba856746de15dae78d4dfR200 / https://docs.gradle.org/3.4/release-notes.html#idea-mapping-has-been-simplified

So, this change is expected, but there might possibly be other side-effects.

I'd be happy to improve the situation, but first: could you try without the `net.ltgt.apt-idea` plugin and see how a standard Gradle configuration works in your case?
You can mimick the configuration for the generated sources folder in IDEA with the following (you might have to manually un-exclude the build directory and re-exclude its subdirectories (with the exception of the one containing the generated sources of course)):
```gradle
idea {
  module {
    sourceDirs += mainSourceSet.output.generatedSourcesDir
    testSourceDirs += testSourceSet.output.generatedSourcesDir
    generatedSourceDirs += [ mainSourceSet.output.generatedSourcesDir, testSourceSet.output.generatedSourcesDir ]
  }
}
```

I'll try to either detect (if possible) whenever we're running in IntelliJ IDEA to configure the project **and** the build will be delegated to Gradle, or add a mean to configure the behavior of the plugin (to only configure source dirs and exclude annotation processors, when you don't want to use annotation processing in IntelliJ IDEA anyway because you're delegating the build to Gradle).

I personally delegate everything to Gradle, including "running" the projects, so I never have that kind of issue.

Do you have a small project I could use to do some testing? (along with instructions how to reproduce the issue in IntelliJ IDEA)

**Comment by mdeinum** [comment]
Without the plugin and with the `eclipse` and `idea` plugins enabled the scope of the dependency (in IntelliJ) is **Compile**. 

Adding the `net.ltgt.apt-idea` and `net.ltgt.apt-eclipse` (we have both eclipse and idea users) and removing the `eclipse` and `idea` plugins still leaves the scope of the dependency to be **Compile**. 

However when using the `net.ltgt.apt-idea` and `net.ltgt.apt-eclipse` plugins _and_ adding dependencies to the `apt` configuration  it suddenly switches to **Provided**.

The application is a [Spring Boot](https://projects.spring.io/spring-boot/) application which from the command line works as expected, but when running in IntelliJ it fails. Running it in the IDE is nice for debugging purposes.

**Comment by tbroyer** [comment]
> However when using the `net.ltgt.apt-idea` and `net.ltgt.apt-eclipse` plugins and adding dependencies to the `apt` configuration it suddenly switches to **Provided**.

OK, so it looks like the IntelliJ IDEA Gradle integration works differently than the Gradle `idea` plugin, and somehow disregards (parts of) its configuration. I'll have a look at it (fortunately IntelliJ IDEA Community Edition is open source, and the Gradle integration is part of it).

> The application is a Spring Boot application which from the command line works as expected, but when running in IntelliJ it fails. Running it in the IDE is nice for debugging purposes.

I don't know Spring Boot (I generally avoid anything Spring) but reading the docs and sources, it looks like `./gradlew bootRun` would accept a `--debug` flag (as the `bootRun` task's type extends Gradle's `JavaExec`), so debugging your app would mean running `./gradlew bootRun --debug` (possibly from within IntelliJ IDEA too) and then using a remote debug configuration in IntelliJ IDEA. Consider this a workaround until maybe I find (time to find) a way to make things work out of the box. (apparently you should also be able to delegate run actions to Gradle, so you could possibly run/debug your main class while using Gradle's dependency management: https://mrhaki.blogspot.fr/2016/11/gradle-goodness-delegate-build-and-run.html)

**Comment by tbroyer** [comment]
This should hopefully be fixed in v0.12.
